File: backend/app/services/guide_service.py
import logging
from ..db.guides_repo import get_guide, save_guide
from .openai_service import fetch_guide_via_search

logger = logging.getLogger(__name__)

APP_KEYWORDS: dict[str, list[str]] = {
    "토스": ["toss", "토스"],
    "카카오뱅크": ["카카오뱅크", "kakaobank"],
    "신한": ["신한", "sol", "쏠"],
    "국민": ["국민", "kb", "스타뱅킹"],
    "하나": ["하나", "하나은행"],
    "우리": ["우리", "우리은행"],
    "농협": ["농협", "nh"],
}

# 앱별 공식 헬프센터 URL
APP_HELP_URLS: dict[str, str] = {
    "토스": "https://help.toss.im",
    "카카오뱅크": "https://www.kakaobank.com/Help",
    "신한": "https://www.shinhan.com/hpe/index.jsp#050101000000",
    "국민": "https://www.kbstar.com/quics?page=C102050",
    "하나": "https://www.kebhana.com/cont/help/index.jsp",
    "우리": "https://www.wooribank.com/wb/DrawWBPHP?pgmId=PCUST0010",
    "농협": "https://www.nonghyup.com/guide/faq.do",
}

TASK_KEYWORDS: dict[str, list[str]] = {
    "계좌이체":    ["계좌이체", "송금", "이체", "보내"],
    "잔액조회":    ["잔액", "조회", "얼마"],
    "공과금납부":  ["공과금", "납부", "요금"],
    "OTP발급":    ["otp", "일회용비밀번호"],
    "계좌개설":    ["계좌개설", "계좌 만들기", "통장개설", "통장 만들기"],
    "이체한도변경": ["이체한도", "송금한도", "한도 변경", "한도변경"],
}


def _detect_app(text: str) -> str | None:
    text_lower = text.lower()
    for app_name, keywords in APP_KEYWORDS.items():
        if any(k in text_lower for k in keywords):
            return app_name
    return None


def _detect_task(text: str) -> str | None:
    text_lower = text.lower()
    for task_name, keywords in TASK_KEYWORDS.items():
        if any(k in text_lower for k in keywords):
            return task_name
    return None


def detect_from_history(messages: list[str]) -> tuple[str, str] | None:
    """대화 이력 전체에서 앱과 작업을 각각 수집해 조합"""
    detected_app = None
    detected_task = None

    # 최신 메시지 우선으로 탐색
    for msg in reversed(messages):
        if not detected_app:
            detected_app = _detect_app(msg)
        if not detected_task:
            detected_task = _detect_task(msg)
        if detected_app and detected_task:
            break

    if detected_app and detected_task:
        return detected_app, detected_task
    return None


async def get_or_fetch_guide(messages: list[str]) -> tuple[str, str | None] | None:
    detected = detect_from_history(messages)
    if not detected:
        logger.info("[가이드] 앱/작업 키워드 감지 안됨 → 일반 응답")
        return None

    app_name, task = detected
    logger.info("[가이드] 감지: 앱=%s, 작업=%s", app_name, task)

    cached = await get_guide(app_name, task)
    if cached:
        content, source_url = cached
        logger.info("[가이드] DB 캐시 히트 → web search 생략 (출처: %s)", source_url)
        return content, source_url

    logger.info("[가이드] 캐시 없음 → web search 시작")
    content, _ = await fetch_guide_via_search(app_name, task)
    source_url = APP_HELP_URLS.get(app_name)
    await save_guide(app_name, task, content, source_url)
    logger.info("[가이드] web search 완료 → DB 저장 (출처: %s)", source_url)
    return content, source_url

File: backend/src/main/java/com/digitalhelper/service/GuideService.java
package com.digitalhelper.service;

import com.digitalhelper.entity.Guide;
import com.digitalhelper.repository.GuideRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
public class GuideService {

    private static final Logger log = LoggerFactory.getLogger(GuideService.class);
    private static final int CACHE_DAYS = 30;

    private static final Map<String, String> APP_HELP_URLS = Map.of(
            "토스",      "https://help.toss.im",
            "카카오뱅크", "https://www.kakaobank.com/Help",
            "신한",      "https://www.shinhan.com/hpe/index.jsp#050101000000",
            "국민",      "https://www.kbstar.com/quics?page=C102050",
            "하나",      "https://www.kebhana.com/cont/help/index.jsp",
            "우리",      "https://www.wooribank.com/wb/DrawWBPHP?pgmId=PCUST0010",
            "농협",      "https://www.nonghyup.com/guide/faq.do"
    );

    private final GuideRepository guideRepository;
    private final OpenAiService openAiService;
    private final KeywordDetector keywordDetector;

    public GuideService(GuideRepository guideRepository, OpenAiService openAiService, KeywordDetector keywordDetector) {
        this.guideRepository = guideRepository;
        this.openAiService = openAiService;
        this.keywordDetector = keywordDetector;
    }

    /** 대화 이력에서 앱/작업 감지 후 가이드 반환. 없으면 null */
    public GuideResult getOrFetchGuide(List<String> userMessages) {
        var detection = keywordDetector.detectFromHistory(userMessages);
        if (!detection.hasBoth()) {
            log.info("[가이드] 앱/작업 키워드 감지 안됨 → 일반 응답");
            return null;
        }

        String appName = detection.app();
        String task = detection.task();
        log.info("[가이드] 감지: 앱={}, 작업={}", appName, task);

        Optional<Guide> cached = guideRepository.findValidGuide(appName, task, LocalDateTime.now());
        if (cached.isPresent()) {
            log.info("[가이드] DB 캐시 히트 → web search 생략");
            return new GuideResult(cached.get().getContent(), cached.get().getSourceUrl());
        }

        log.info("[가이드] 캐시 없음 → web search 시작");
        String content = openAiService.fetchGuideViaSearch(appName, task);
        String sourceUrl = APP_HELP_URLS.get(appName);

        saveOrUpdate(appName, task, content, sourceUrl);
        log.info("[가이드] web search 완료 → DB 저장 (출처: {})", sourceUrl);
        return new GuideResult(content, sourceUrl);
    }

    private void saveOrUpdate(String appName, String task, String content, String sourceUrl) {
        LocalDateTime expiresAt = LocalDateTime.now().plusDays(CACHE_DAYS);
        Guide guide = guideRepository.findByAppNameAndTask(appName, task)
                .orElse(new Guide(appName, task, content, sourceUrl, expiresAt));
        guide.setContent(content);
        guide.setSourceUrl(sourceUrl);
        guide.setCreatedAt(LocalDateTime.now());
        guide.setExpiresAt(expiresAt);
        guideRepository.save(guide);
    }

    public record GuideResult(String content, String sourceUrl) {}
}
